### ISSUE TITLE
Need flutter_tool support for adding new device types as add-ons

### ISSUE THREAD

**Issue by JakeSays** [flutter/flutter#33736]
We need the ability to create a custom flutter sdk that uses our custom flutter engine. Currently we must build flutter apps using the command line and the `--local-engine` flutter switch. This is cumbersome and cannot be integrated in to the tooling (vscode, etc). At least I cannot find a way to do it.

From looking at the ide plugins it appears there is no way to support anything but android and ios sdks.

**Comment by stuartmorgan** [comment]
Can you elaborate on what it is exactly that you want to be able to do? "create a custom flutter sdk" sounds like forking Flutter and modifying it; what is preventing you from being able to do that?

Do you actually mean you want to have a *stock* Flutter SDK that always uses a custom engine? If that's the case, it seems like --local-engine does what you want, and feature request for your IDE of choice to support setting that flag is the missing piece, but maybe I'm not understanding.

**Comment by JakeSays** [comment]
Yes, a stock flutter sdk with a custom engine is what I mean. The problem I'm having is --local-engine also requires a parallel engine source/build tree, complete with git repos, not just the engine binary itself. We currently create a tarball out of both the flutter and engine repos, but it's 12GB in size.

As I understand it there is some form of libflutter_engine.so that is contained in the SDK, one for each platform I think, and compatible host gen_snapshot.

What I would like is to have a flutter sdk "installer" that I can reference in vscode that has enough in it to support building and deploying an app to an embedded device.

Also I noticed that the sdk contains target binaries for only android and ios, which is another issue altogether.

**Comment by stuartmorgan** [comment]
> which is another issue altogether.

Is it? It sounds like you are not just trying to use a local engine build, but also target a new type of device; that's going to require a lot more than just something `--local-engine`-like.

If the goal is in fact to get a setup where someone can take a stock Flutter tree and plugin in all of the pieces necessary for an entirely new device pipeline (new engine build, new embedding, new device types, new runner templates, etc.) it would probably be better to provide an overall description of what you want to be able to build so that can be evaluated holistically rather than individual requests.

**Comment by JakeSays** [comment]
Yes, exactly - I need to target a new class of devices, specifically embedded devices that are based on dragonboard, jetson nano, raspberry pi, etc. Basically full screen kiosk style scenarios. I also have handheld point of sale terminals that I need to support.

This is what I meant by the sdk targeting only android/ios. I downloaded the 1.5.4 sdk and below is the layout.

Each directory under bin/cache/artifacts/engine contains a flutter.jar file that contains a platform specific libflutter.so. Some directories also contain gen_snapshot's. I am not sure why IOS isn't in there, but I suspect it's because I'm not on a mac.

So basically what I need (I *think*) is an sdk that has a directory for my custom embedder. 

My current thinking is I may have to derive my own vs code extension from the official one that handles all of this.

I've been focused on getting the embedders up and running. I currently have embedders for a software framebuffer, raspberry pi native, and SDL, and I will also have kmsdrm and wayland support when I'm done.

Now I need to create tooling support for them.

![flutter-sdk-layout](https://user-images.githubusercontent.com/49178478/58775859-d3f90380-8584-11e9-8a4d-9a17ffca03b6.png)

**Comment by stuartmorgan** [comment]
> So basically what I need (I think) is an sdk that has a directory for my custom embedder.

You would need a lot more than that; you'd need changes throughout `flutter_tool` to understand the new device type.

And the feature you described in your initial request (something that behaves like `--local-engine`) wouldn't actually help you, because that is for replacing one of the existing libraries with a locally built one, which doesn't seem to be what you want at all.

> My current thinking is I may have to derive my own vs code extension from the official one

If you are going that route, presumably you wouldn't need any changes to the Flutter tooling.

It seems like you should probably spend more time exploring the `flutter_tool` and Dart Code sources to get a better understanding of what would be involved in each approach before filing requests, since it doesn't sound like you've even decided on whether you want to use the stock Flutter tooling at all.

**Comment by stuartmorgan** [comment]
Retitling to more accurately describe what you would actually need to support a new device type without forking Flutter.

**Comment by jonahwilliams** [comment]
This is being worked on, through it is a significant change in the tool and there is substantially more to do.

The tracking issue is here: https://github.com/flutter/flutter/issues/25377

If you're interested in trying it out when I have a preview ready please let me know

**Comment by JakeSays** [comment]
@jonahwilliams Yes I am very interested! Just let me know when you have something. Thanks!

**Comment by JakeSays** [comment]
@jonahwilliams Hey have you made any progress on this?
